wallet/src/sender.rs

```diff
@@ -118,19 +120,34 @@ pub fn issue_burn_tx(
 	})?
 }
 
-fn inputs_and_change(coins: &Vec<OutputData>, keychain: &Keychain, fingerprint: Fingerprint, wallet_data: &mut WalletData, amount: u64) -> Result<Vec<Box<build::Append>>, Error> {
+fn inputs_and_change(
+	coins: &Vec<OutputData>,
+	keychain: &Keychain,
+	root_key_id: Identifier,
+	wallet_data: &mut WalletData,
+	amount: u64,
+) -> Result<Vec<Box<build::Append>>, Error> {
 
 	let mut parts = vec![];
 
-	// calculate the total in inputs, fees and how much is left
+	// calculate the total across all inputs, and how much is left
 	let total: u64 = coins.iter().map(|c| c.value).sum();
-	let fee = tx_fee(coins.len(), 2, None);
-	let shortage = (total as i64) - (amount as i64) - (fee as i64);
+	let shortage = (total as i64) - (amount as i64);
 	if shortage < 0 {
 		return Err(Error::NotEnoughFunds((-shortage) as u64));
 	}
+
+	// sender is responsible for setting the fee on the partial tx
+	// recipient should double check the fee calculation and not blindly trust the
+	// sender
+	let fee = tx_fee(coins.len(), 2, None);
 	parts.push(build::with_fee(fee));
-	let change = total - amount - fee;
+
+	// if we are spending 10,000 coins to send 1,000 then our change will be 9,000
+	// the fee will come out of the amount itself
+	// if the fee is 80 then the recipient will only receive 920
+	// but our change will still be 9,000
+	let change = total - amount;
```

Comment:
I think a bug was introduced here when the fee calc was put in place.
The fee comes out of the amount, not in addition to it.
This was producing an invalid tx as the amounts were no longer summing to 0 correctly.

Reply:
Ah yes, good catch, sorry for the oversight. Maybe some more wallet tests would be good? I can create a github issue for it.